File: LeetCode/Tag/Hash-Table/java/1679.max-number-of-k-sum-pairs.java
class Solution {
    public int maxOperations(int[] nums, int k) {
        HashMap<Integer, Integer> freq = new HashMap<>();
        int count = 0;
        for (int x : nums) {
            if (freq.getOrDefault(k-x, 0) > 0) {
                ++count;
                freq.put(k-x, freq.get(k-x) - 1);
            } else {
                freq.put(x, freq.getOrDefault(x, 0) + 1);
            }
        }
        return count;
    }
}

File: LeetCode/Tag/Hash-Table/cpp/1679.max-number-of-k-sum-pairs.cpp
class Solution {
public:
    int maxOperations(vector<int>& nums, int k) {
        unordered_map<int, int> freq;
        int count = 0, minFreq;
        for (int x : nums ) {
            ++freq[x];
            if (2*x == k) {
                count += freq[x]/2;
                freq[x] %= 2;
            } else if (freq[k-x] > 0) {
                minFreq = min(freq[x], freq[k-x]);
                freq[x] -= minFreq;
                freq[k-x] -= minFreq;
                count += minFreq;
            }
        }
        return count;
    }
};